libbeat/common/kubernetes/watcher.go

```diff
@@ -0,0 +1,261 @@
+package kubernetes
+
+import (
+	"context"
+	"errors"
+	"sync"
+	"time"
+
+	"github.com/elastic/beats/libbeat/common/bus"
+	"github.com/elastic/beats/libbeat/logp"
+
+	"github.com/ericchiang/k8s"
+	corev1 "github.com/ericchiang/k8s/api/v1"
+)
+
+// Watcher reads Kubernetes events and keeps a list of known pods
+type Watcher interface {
+	// Start watching Kubernetes API for new containers
+	Start() error
+
+	// Stop watching Kubernetes API for new containers
+	Stop()
+
+	// ListenStart returns a bus listener to receive pod started events, with a `pod` key holding it
+	ListenStart() bus.Listener
+
+	// ListenUpdate returns a bus listener to receive pod updated events, with a `pod` key holding it
+	ListenUpdate() bus.Listener
+
+	// ListenStop returns a bus listener to receive pod stopped events, with a `pod` key holding it
+	ListenStop() bus.Listener
+}
+
+type podWatcher struct {
+	sync.RWMutex
+	client              Client
+	syncPeriod          time.Duration
+	cleanupTimeout      time.Duration
+	nodeFilter          k8s.Option
+	lastResourceVersion string
+	ctx                 context.Context
+	stop                context.CancelFunc
+	bus                 bus.Bus
+	pods                map[string]*Pod      // pod id -> Pod
+	deleted             map[string]time.Time // deleted annotations key -> last access time
+}
+
+// Client for Kubernetes interface
+type Client interface {
+	ListPods(ctx context.Context, namespace string, options ...k8s.Option) (*corev1.PodList, error)
+	WatchPods(ctx context.Context, namespace string, options ...k8s.Option) (*k8s.CoreV1PodWatcher, error)
+}
+
+// NewWatcher initializes the watcher client to provide a local state of
+// pods from the cluster (filtered to the given host)
+func NewWatcher(client Client, syncPeriod, cleanupTimeout time.Duration, host string) Watcher {
+	ctx, cancel := context.WithCancel(context.Background())
+	return &podWatcher{
+		client:              client,
+		cleanupTimeout:      cleanupTimeout,
+		syncPeriod:          syncPeriod,
+		nodeFilter:          k8s.QueryParam("fieldSelector", "spec.nodeName="+host),
+		lastResourceVersion: "0",
+		ctx:                 ctx,
+		stop:                cancel,
+		pods:                make(map[string]*Pod),
+		deleted:             make(map[string]time.Time),
+		bus:                 bus.New("kubernetes"),
+	}
+}
+
+func (p *podWatcher) syncPods() error {
+	logp.Info("kubernetes: %s", "Performing a pod sync")
```

Comment:
Perhaps we could use a kubernetes prefix for the above log message (see comment).